src/net.cpp

```diff
@@ -2917,20 +2917,21 @@ bool CConnman::ForNode(NodeId id, std::function<bool(CNode* pnode)> func)
     return found != nullptr && NodeFullyConnected(found) && func(found);
 }
 
-int64_t CConnman::PoissonNextSendInbound(int64_t now, int average_interval_seconds)
+std::chrono::microseconds CConnman::PoissonNextSendInbound(std::chrono::microseconds now, std::chrono::microseconds average_interval)
 {
-    if (m_next_send_inv_to_incoming < now) {
+    if (m_next_send_inv_to_incoming.load() < now) {
         // If this function were called from multiple threads simultaneously
         // it would possible that both update the next send variable, and return a different result to their caller.
         // This is not possible in practice as only the net processing thread invokes this function.
-        m_next_send_inv_to_incoming = PoissonNextSend(now, average_interval_seconds);
+        m_next_send_inv_to_incoming = PoissonNextSend(now, average_interval);
     }
     return m_next_send_inv_to_incoming;
 }
 
-int64_t PoissonNextSend(int64_t now, int average_interval_seconds)
+std::chrono::microseconds PoissonNextSend(std::chrono::microseconds now, std::chrono::microseconds average_interval)
 {
-    return now + (int64_t)(log1p(GetRand(1ULL << 48) * -0.0000000000000035527136788 /* -1/2^48 */) * average_interval_seconds * -1000000.0 + 0.5);
+    double unscaled = -log1p(GetRand(1ULL << 48) * -0.0000000000000035527136788 /* -1/2^48 */);
+    return now + std::chrono::round<std::chrono::microseconds>(unscaled * average_interval);
```

Comment:
It does seem better to have the average interval in seconds. Updated.

I felt that the code reads cleaner with `chrono::round` and have left it in for now. If other reviewers also think more optimal assembly is necessary here, I'll be happy to change it.

Reply:
Optimal assembly doesn't matter so much (we're calculating log1p at the same time which presumably isn't that quick), it's more that it's not particularly obvious to me that using `chrono::round` doesn't introduce a bug, and I don't see any need for the change. Obviously if it does introduce a bug, that would be the compiler's fault rather than yours per se, but that happens too.

I think an alternative way of writing what we currently do would be:

```c++
    constexpr auto HALFMICRO = 0.5 * 1us;
    double unscaled = -log1p(GetRand(1ULL << 48) * -0.0000000000000035527136788 /* -1/2^48 */);
    return now + std::chrono::duration_cast<std::chrono::microseconds>(unscaled * average_interval + HALFMICRO);
```

It doesn't generate precisely the same asm as the current code because the "-" isn't attached to the factor of 1M anymore but I think that's the only difference, so it seems easier to be sure that's actually not changing any behaviour.

(I don't think there's any real value in rounding to the nearest microsecond rather than just truncating -- everything we do ends up being at intervals of 50,000 or 100,000 microseconds anyway)

Reply:
There's definitely a tradeoff between the probability of introducing a bug and modernizing our codebase. When we choose to embrace a new stdlib or language feature, I'd propose we should do it _relatively_ consistently (lower level code will be more resistant). It seems we're adopting `std::chrono`, so `std::chrono::round` is not that different.

Reply:
I think if we can't trust the compiler to produce correct code, and can't trust our test to catch the issue either, there is nothing we can do preemptively other than write more tests. Choosing one version of the code over another in the absence of any known compiler miscompilation bugs doesn't seem too helpful to me, because we don't know if any of our dependencies use that piece of code in either a header or library.

(I am not saying which version to pick here. I am happy to review and ACK any version that does the right thing)

Reply:
> It seems we're adopting `std::chrono`, so `std::chrono::round` is not that different.

I just explained why std::chrono::round is different -- it does different, substantially more complicated logic than the current code [0]. Converting to `std::chrono::duration` doesn't do different logic to just having an `int64_t`, it just adds type safety to help avoid mixing different scales. We don't need to use `::round` to use `std::chrono`-- we can do `duration_cast` to truncate instead. Rounding to even has its place (https://mathematica.stackexchange.com/questions/2116/why-round-to-even-integers) but none of those reasons apply here.

[0] It's kind-of weird that it does more logic -- round to even is a standard IEEE754 rounding mode so it ought not to require custom template code or result in complicated assembly.

> Choosing one version of the code over another in the absence of any known compiler miscompilation bugs

In the absence of a good reason to make a change, don't make a change. If there's two approaches that achieve the same result, choose the simpler one.

Reply:
I have now actually benchmarked the `PoissonNextSend` in 3 different variants: original, round, duration_cast. Here are the results:

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|            2,338.00 |          427,716.00 |    0.4% |        6,721.29 |        7,467.00 |  0.900 |         193.00 |    0.5% |      0.00 | `old`
|            2,380.33 |          420,109.23 |    1.8% |        6,749.57 |        7,603.17 |  0.888 |         199.29 |    0.4% |      0.00 | `round`
|            2,338.12 |          427,693.13 |    0.3% |        6,749.00 |        7,486.00 |  0.902 |         199.29 |    0.4% |      0.00 | `duration_cast`

The runtime difference here umeasurable. The overhead is completely dwarfed by `GetRand`. When I then replaced `GetRand` with the simplest possible random generator I could think of:

```cpp
inline uint64_t myRand(uint64_t maxnum) {
    static uint64_t x = 1;
    auto tmp = x % maxnum;
    x *= 123123121;
    return tmp;
}
```

Ran the benchmark again:

|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|               18.43 |       54,269,715.35 |    0.4% |          110.27 |           58.90 |  1.872 |          13.78 |    4.5% |      0.00 | `original`
|               24.60 |       40,642,221.28 |    0.3% |          143.80 |           78.60 |  1.830 |          20.85 |    3.1% |      0.00 | `round`
|               24.55 |       40,724,958.20 |    0.3% |          143.82 |           78.57 |  1.831 |          20.85 |    3.0% |      0.00 | `duration_cast`

There is a slight difference between `duration_cast` and `round`, but hardly significant. 

So at least on my Intel CPU there is no performance argument for duration_cast in this case.

Reply:
With no material performance difference, my preference aside, I do not see a reason to hold up this PR. Updated the code to use `duration_cast`.

Reply:
> In the absence of a good reason to make a change, don't make a change. If there's two approaches that achieve the same result, choose the simpler one.

Obviously I agree with that statement, but it doesn't seem applicable here. First, it doesn't look like the duration cast produces the same binary for me: https://godbolt.org/z/nejf46. And second, the `round` variant is simpler and shorter code.

Reply:
I believe the differences are that `average_interval` goes from `int` to `int64_t` precision, and `0.5us` is a long double while `0.5` is a double. I think replacing `chrono::round<chrono::microseconds>` with:

```c++
template<typename Dur, typename Dur2>
static inline Dur chrono_round_half_up(Dur2 dur)
{
    return std::chrono::duration_cast<Dur>(Dur{1}*0.5 + dur);
}
```

will generate the same code that we currently get on clang (except for `average_interval_seconds` being int vs int64_t):  https://godbolt.org/z/rxjdc8 . gcc seems to generate different code due to negating the log1p result then multiplying by 1M rather than just multiplying by -1M.